game/songitems.cc

```diff
@@ -71,32 +68,27 @@ void SongItems::addSong(std::shared_ptr<Song> song) {
 	m_songs.insert(si);
 }
 
-int SongItems::lookup(std::shared_ptr<Song> song) const {
-	for (auto const& s: m_songs) {
-		if (song->collateByArtistOnly == s.artist && song->collateByTitleOnly == s.title) return s.id;
-	}
-	return -1;
+SongId SongItems::lookup(Song const& song) const {
+	auto const& si = std::find_if(m_songs.begin(), m_songs.end(), [song](SongItem const& si) {
+		return UnicodeUtil::toLower(song.collateByArtistOnly) == UnicodeUtil::toLower(si.artist) && UnicodeUtil::toLower(song.collateByTitleOnly) == UnicodeUtil::toLower(si.title);
+	});
+	if (si != m_songs.end()) return si->id;
+	return std::nullopt;
 }
 
-int SongItems::lookup(Song& song) const {
-	for (auto const& s: m_songs) {
-		if (song.collateByArtistOnly == s.artist && song.collateByTitleOnly == s.title) return s.id;
-	}
-	return -1;
-}
-
-std::string SongItems::lookup(int id) const {
+std::string SongItems::lookup(SongId id) const {
 	SongItem si;
-	si.id = id;
+	if (!id) return "Invalid song ID";
+	si.id = id.value();
 	auto it = m_songs.find(si);
 	if (it == m_songs.end()) return "Unknown Song";
 	else if (!it->song) return it->artist + " - " + it->title;
 	else return it->song->artist + " - " + it->song->title;
 }
 
-int SongItems::assign_id_internal() const {
+unsigned SongItems::assign_id_internal() const {
 	// use the last one with highest id
 	auto it = m_songs.rbegin();
 	if (it != m_songs.rend()) return it->id+1;
-	else return 1; // empty set
```

Comment:
you change 1 to 0 on purpose?

Reply:
I did, I figured breakage potential for existing users is very small and starting at 1 is weird anyway.